### DISCUSSION TITLE
Improve the way to write bindings

### DISCUSSION THREAD

**Discussion by craff** [ocaml/ocaml#12471] [Features]
Following the PR #12375, I wish to work more on improving the C binding. Gaining both efficiency and reliability. 

Here are a few ideas:

FLAGS (issue #12366)
-------------------------------

These are very frequent in C library. I do not like the list approach (allocates and slow) I see two directions:

- using variants (With the PR #12375, they may be untagged): it requires a way to choose the representation of contants such as
  ```
  type a = A [@eq 0x1] | B [@eq 0x2] | C [@eq 0x4]
  ```
  it also requires an efficient identity function to int to be able to use lor, which could be obtained by optimizing the pattern matching (it already works, we just have to adapt it to [@eq]. A second solution it to accept (x :> int) for all variant type with only constants (or for all immediate type. 

@gasche and @yallop in the above mentioned issue prefer the first solution. I prefer the second, because for OpenGL/EGL, writing the function is really painfull as there are many contants.  A mixed approach is a ppx, but I owuld like support in the base compiler. 

- using integer (private or abstract). This already works, but we can not pattern match on constants. This is not often needed, but pattern matching on compile time constants would be nice. 

Remark: conversion of immediate types to int allows for indexing an array.

Unix.file_descr
----------------------

These are not recognize as immediate types because they are not integer on windows. A function doing the conversion to int in both direction would be nice. It also allows for sending file_descr over channel as 8 or even 4 bytes instead of 21 with Marshall.

String and bytes as arguments
---------------------------------------------

They do not require any flags, as they are represented by a C compatible pointer. But if one look at write/read and many other function, an offset would be nice. example
```
external f : string -> int [@offset] -> unit = ...
```

Custom blocks (including bigarrays), as arguments
--------------------------------------------------------------------------

Would be nice to allow to use [@unbox] on them (or create a specific tag)

Verification
----------------

There is no check that types agree between the external declaration and the C implementation. For a language that is "strongly statistycally typed" this sounds like a problem. I see two ways

- checking a .h file from OCaml (there are many C related tools written in OCaml to do that job) 
- generating some C file from OCaml that compiles only if there is no problem. 

The second approach is simpler and has no dependencies. @yallop [ctypes](https://github.com/yallop/ocaml-ctypes) does that job, could be integrated in the compiler ?

More int size available
--------------------------------

Currently, untagged should corresponds to natint. I would be nice if char/short, etc would be available, and if OCaml generates some C file with some ##if directive, it seems feasible. This also requires more code to emit the call to C

**Comment by gasche** [comment]
> There is no check that types agree between the external declaration and the C implementation

Please look at this work from @edwintorok :

https://arxiv.org/abs/2307.14909

>  Migration to OCaml 5 requires updating a lot of C bindings due to the removal of naked pointer support. Writing OCaml user-defined primitives in C is a necessity, but is unsafe and error-prone. It does not benefit from either OCaml's or C's type checking, and existing C static analysers are not aware of the OCaml GC safety rules, and cannot infer them from existing macros alone.The alternative is automatically generating C stubs, which requires correctly managing value lifetimes. Having a static analyser for OCaml to C interfaces is useful outside the OCaml 5 porting effort too.
After some motivating examples of real bugs in C bindings a static analyser is presented that finds these known classes of bugs. The tool works on the OCaml abstract parse and typed trees, and generates a header file and a caller model. Together with a simplified model of the OCaml runtime this is used as input to a static analysis framework, Goblint. An analysis is developed that tracks dereferences of OCaml values, and together with the existing framework reports incorrect dereferences. An example is shown how to extend the analysis to cover more safety properties.
The tools and runtime models are generic and could be reused with other static analysis tools. 

This is of general interest, but in particular it has a tool building C files from OCaml external declarations, to have a C compiler check the consistency with the actual C implementations.

**Comment by gasche** [comment]
> It would be nice to allow to use `@unbox` on [custom blocks]

One difficulty here is the lifetime of the underlying OCaml values. Normally external calls do not keep their value arguments alive, it is the callee's responsibility to root the values if they are still useful. For unboxed numeric arguments, the OCaml-side values are discarded and may be collected during the external call computation. For custom values this approach would be incorrect, as this means that the GC could call the custom finalizer in the middle of the C function call.

The only approach I see is to decide that custom blocks should remain alive for the whole duration of the external call, artificially extending the duration compared to normal function arguments. This requires low-level changes to parts of the compiler I don't know about (the static computation of local roots). I am hesitant about a mostly-cosmetic feature requiring such ad-hoc changes to the compiler.

(Thinking about this more: we cannot easily do this statically when computing local roots, because at that point we may not know which values will dynamically receive custom blocks. We could either use a separate unboxing annotation for this -- that extends the lifetime of the OCaml block -- or insert a dynamic check for custom blocks to be rooted dynamically. The separate unboxing annotation feels more reasonable, but it means that users who forget to use it will get random crashes in their programs.)
